File: bfs/coin_change_2.rb
require 'set'

# TLE Solution, bcuz only counts are needed. Not the actual paths.
# Also, hacky in the sense of eliminating duplicate paths using sort on curr_path @$$)(Q@)#(*#)
# @param {Integer[]} coins
# @param {Integer} amount
# @return {Integer}
def change(amount, coins)
    return 0 if amount.zero?
    paths = []

    q, visited = [[0, []]], Set.new([])
    while !q.empty?
        node, path = q.shift
        coins.each do |coin|
            nxt_node = coin + node
            curr_path = path + [coin]

            if nxt_node == amount
                puts "nxt_node: #{nxt_node}, path: #{curr_path}, sorted path: #{curr_path.sort}, visited: #{visited}"
                paths.push(curr_path.sort) if !visited.include?(curr_path.sort)
                visited.add(curr_path.sort) # Only way to keep 
                next
            end
            next if visited.include?(curr_path.sort)
            next if nxt_node > amount

            visited.add(curr_path.sort)
            q.push([nxt_node, curr_path])
        end
    end

    p paths
    paths.count
end


# 322. Coin Change
# https://leetcode.com/problems/coin-change/

require 'test/unit'
extend Test::Unit::Assertions

assert_equal(change(5, [1,2, 5]), 4)

File: sandbox/coin_change_2.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)
        dp = [0] * (amount + 1)
        dp[0] = 1

        for c in coins:
            for j in range(1, amount + 1):
                if c > j: continue
                dp[j] += dp[j - c]
        return dp[amount]


# 518. Coin Change 2
# https://leetcode.com/problems/coin-change-2/

# dp[i] = no. of ways to form amount i with all coins
# dp[i] = dp[i] + d[i-c] for every coin value 'c'
# dp[0] = 1
